### PR TITLE
W-14572268: Remove 2 step compilation

### PR DESCRIPTION
**Pull request by elrodro83**
Includes W-14572258: Remove mrjars

### CODE REVIEW

**Review by rbourbonnavarro — CHANGES_REQUESTED** [review]

File: modules/jar-handling-utils/src/main/java/org/mule/runtime/module/artifact/api/classloader/jar/CachingURLStreamHandlerFactory.java [hunk 1]
```diff
@@ -17,7 +17,7 @@
 public class CachingURLStreamHandlerFactory {
 
   public static URLStreamHandlerFactory getCachingURLStreamHandlerFactory() {
-    return new NonCachingURLStreamHandlerFactory();
+    return null;
```
Comment (rbourbonnavarro) [inline modules/jar-handling-utils/src/main/java/org/mule/runtime/module/artifact/api/classloader/jar/CachingURLStreamHandlerFactory.java:20]: We should remove this and just use `null` where appropriate I think.

**Review by rbourbonnavarro — APPROVED** [review]

File: modules/container/src/main/java/org/mule/runtime/container/internal/ContainerModulesDiscovererProvider.java [hunk 2]
```diff
@@ -6,23 +6,32 @@
...
 
   public static ModuleDiscoverer containerModulesDiscoverer() {
-    LOGGER.debug("Default 'ContainerModulesDiscovererProvider' implementation, using 'ClasspathModuleDiscoverer'...");
-    return new ClasspathModuleDiscoverer();
+    if (classloaderContainerJpmsModuleLayer()) {
+      LOGGER.debug("MRJAR 'ContainerModulesDiscovererProvider' implementation, using 'JpmsModuleLayerModuleDiscoverer'...");
+      return new JpmsModuleLayerModuleDiscoverer(new ClasspathModuleDiscoverer());
+    } else {
+      LOGGER.debug("MRJAR 'ContainerModulesDiscovererProvider' implementation, using 'ClasspathModuleDiscoverer'...");
+      return new ClasspathModuleDiscoverer();
```
[the remark this answers is not in the document]
Comment (elrodro83) [reply modules/container/src/main/java/org/mule/runtime/container/internal/ContainerModulesDiscovererProvider.java:33]: proper cleanup will be tackled separately

File: modules/service/src/main/java/org/mule/runtime/module/service/api/artifact/ServiceClassLoaderFactoryProvider.java [hunk 3]
```diff
@@ -27,12 +30,23 @@ public static ClassLoaderConfigurationLoader serviceClassLoaderConfigurationLoad
...
+    if (classloaderContainerJpmsModuleLayer()
+        // Only if the container is modularized it makes sense to load services as module layers
+        && withinModularizedContainer) {
+      LOGGER.debug("MRJAR 'ServiceClassLoaderFactoryProvider' implementation, using 'ServiceModuleLayerFactory'...");
+      final ServiceModuleLayerFactory serviceModuleLayerFactory = new ServiceModuleLayerFactory();
+      serviceModuleLayerFactory.setParentLayerFrom(ServiceClassLoaderFactoryProvider.class);
+      return serviceModuleLayerFactory;
+    } else {
+      LOGGER.debug("MRJAR 'ServiceClassLoaderFactoryProvider' implementation, using 'ServiceClassLoaderFactory'...");
+      return new ServiceClassLoaderFactory();
```
[the remark this answers is not in the document]
Comment (elrodro83) [reply modules/service/src/main/java/org/mule/runtime/module/service/api/artifact/ServiceClassLoaderFactoryProvider.java:45]: proper cleanup will be tackled separately

File: modules/jar-handling-utils/src/main/java/org/mule/runtime/module/artifact/api/classloader/jar/CachingURLStreamHandlerFactory.java (hunk 1 above)
Comment (elrodro83) [reply modules/jar-handling-utils/src/main/java/org/mule/runtime/module/artifact/api/classloader/jar/CachingURLStreamHandlerFactory.java:20]: proper cleanup will be tackled separately

**Review by rbourbonnavarro — DISMISSED** [review]

**Review by rbourbonnavarro — APPROVED** [review]

**Pull request merged**